File: solutions/0530. Minimum Absolute Difference in BST/0530.cpp
// Time complexity: O(n)
// Space complexity: O(n)
class Solution {
public:
    int minDifference = INT_MAX;
    TreeNode* prevNode;

    void inorderTraversal(TreeNode* node) {
        if (node == nullptr) {
            return;
        }

        inorderTraversal(node->left);
        if (prevNode != nullptr) {
            minDifference = min(minDifference, node->val - prevNode->val);
        }
        prevNode = node;
        inorderTraversal(node->right);
    }

    int getMinimumDifference(TreeNode* root) {
        inorderTraversal(root);
        return minDifference;
    }
};

File: solutions/0530. Minimum Absolute Difference in BST/0530.py
# Time complexity: O(n)
# Space complexity: O(n)
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        self.min_diff = float('inf')
        self.prev = None
        
        def inOrderTraversal(node):
            if node is None:
                return
            
            inOrderTraversal(node.left)
            
            if self.prev is not None:
                self.min_diff = min(self.min_diff, node.val - self.prev.val)
            
            self.prev = node
            
            inOrderTraversal(node.right)
        
        inOrderTraversal(root)
        return self.min_diff